pkg/addons/addons.go

```diff
@@ -85,34 +68,103 @@ func readDB(cs *acsapi.ContainerService, c *config.Config) (map[string]unstructu
 	return db, nil
 }
 
-// getDynamicClient returns the server API group resource information and a
-// dynamic client pool.
-func getDynamicClient(cli *discovery.DiscoveryClient) (dynamic.ClientPool, []*discovery.APIGroupResources, error) {
-	grs, err := discovery.GetAPIGroupResources(cli)
-	if err != nil {
-		return nil, nil, err
+// syncWorkloadsConfig iterates over all workload controllers (deployments,
+// daemonsets, statefulsets), walks their volumes, and updates their pod
+// templates with annotations that include the hashes of the content for
+// each configmap or secret.
+func syncWorkloadsConfig(db map[string]unstructured.Unstructured) error {
+	// map config resources to their hashed content
+	configToHash := make(map[string][]byte)
+	for _, o := range db {
+		if o.GroupVersionKind().Kind != "Secret" &&
+			o.GroupVersionKind().Kind != "ConfigMap" {
+			continue
+		}
+
+		dataHasher := sha256.New()
+		for _, v := range jsonpath.MustCompile("$.data").Get(o.Object) {
+			fmt.Fprintf(dataHasher, "%#v", v)
+		}
+		if dataHasher.Size() > 0 {
+			configToHash[KeyFunc(o.GroupVersionKind().GroupKind(), o.GetNamespace(), o.GetName())] = dataHasher.Sum(nil)
+		}
+
+		stringDataHasher := sha256.New()
+		for _, v := range jsonpath.MustCompile("$.stringData").Get(o.Object) {
+			fmt.Fprintf(stringDataHasher, "%#v", v)
+		}
+		if stringDataHasher.Size() > 0 {
+			configToHash[KeyFunc(o.GroupVersionKind().GroupKind(), o.GetNamespace(), o.GetName())] = stringDataHasher.Sum(nil)
+		}
 	}
 
-	rm := discovery.NewRESTMapper(grs, meta.InterfacesForUnstructured)
-	dyn := dynamic.NewClientPool(restconfig, rm, dynamic.LegacyAPIPathResolverFunc)
+	secretGk := schema.GroupKind{Kind: "Secret"}
+	configMapGk := schema.GroupKind{Kind: "ConfigMap"}
+
+	// iterate over all workload controllers and add annotations with the hashes
+	// of every config map or secret appropriately to force redeployments on config
+	// updates.
+	for _, o := range db {
+		if o.GroupVersionKind().Kind != "DaemonSet" &&
+			o.GroupVersionKind().Kind != "Deployment" &&
+			o.GroupVersionKind().Kind != "StatefulSet" {
+			continue
+		}
+
+		volumes := jsonpath.MustCompile("$.spec.template.spec.volumes.*").MustGetObject(o.Object)
 
-	return dyn, grs, nil
+		if secretData, found := volumes["secret"]; found {
+			secretName := jsonpath.MustCompile("$.secretName").MustGetString(secretData)
+			key := fmt.Sprintf("checksum/secret-%s", secretName)
+			secretKey := KeyFunc(secretGk, o.GetNamespace(), secretName)
+			setPodTemplateAnnotation(key, string(configToHash[secretKey]), o)
```

Comment:
does this need to be
```
if hash, found := configToHash[secretKey]; found {
	setPodTemplateAnnotation(key, base64.StdEncoding.EncodeToString(hash), o)
}
```
ie 1- don't record annotations for secrets we don't know about (like serving certs)
2- base64 the hash

Reply:
Ignoring serving certs is fine by me, not sure the extra base64 step is needed.